staging/src/k8s.io/component-base/metrics/legacyregistry/registry.go

```diff
@@ -17,112 +17,53 @@ limitations under the License.
-	defer globalRegistryFactory.registrationLock.Unlock()
-	if globalRegistryFactory.kubeVersion != nil {
-		if globalRegistryFactory.kubeVersion.String() != ver.String() {
-			panic(fmt.Sprintf("Cannot load a global registry more than once, had %s tried to load %s",
-				globalRegistryFactory.kubeVersion.String(),
-				ver.String()))
-		}
-		return nil
-	}
-	registrationErrs := make([]error, 0)
-	preloadedMetrics := []prometheus.Collector{
-		prometheus.NewGoCollector(),
-		prometheus.NewProcessCollector(prometheus.ProcessCollectorOpts{}),
-	}
-	globalRegistryFactory.globalRegistry = metrics.NewKubeRegistry(ver)
-	globalRegistryFactory.globalRegistry.RawMustRegister(preloadedMetrics...)
-	globalRegistryFactory.kubeVersion = &ver
-	for _, c := range globalRegistryFactory.registerQueue {
-		err := globalRegistryFactory.globalRegistry.Register(c)
-		if err != nil {
-			registrationErrs = append(registrationErrs, err)
-		}
-	}
-	for _, c := range globalRegistryFactory.mustRegisterQueue {
-		globalRegistryFactory.globalRegistry.MustRegister(c)
-	}
-	return registrationErrs
-}
-
-// Register registers a collectable metric, but it uses a global registry. Registration is deferred
-// until the global registry has a version to use.
+// Register registers a collectable metric but uses the global registry
 func Register(c metrics.Registerable) error {
-	globalRegistryFactory.registrationLock.Lock()
-	defer globalRegistryFactory.registrationLock.Unlock()
-
-	if reflect.DeepEqual(globalRegistryFactory.globalRegistry, noopRegistry{}) {
-		globalRegistryFactory.registerQueue = append(globalRegistryFactory.registerQueue, c)
-		return nil
-	}
-
-	return globalRegistryFactory.globalRegistry.Register(c)
+	return defaultRegistry.Register(c)
 }
 
-// MustRegister works like Register but registers any number of
-// Collectors and panics upon the first registration that causes an
-// error. Registration is deferred  until the global registry has a version to use.
+// MustRegister registers registerable metrics but uses the global registry.
 func MustRegister(cs ...metrics.Registerable) {
-	globalRegistryFactory.registrationLock.Lock()
-	defer globalRegistryFactory.registrationLock.Unlock()
+	defaultRegistry.MustRegister(cs...)
+}
+
+// RawMustRegister registers prometheus collectors but uses the global registry, this
+// bypasses the metric stability framework
+//
+// Deprecated
+func RawMustRegister(cs ...prometheus.Collector) {
+	defaultRegistry.RawMustRegister(cs...)
+}
 
-	if reflect.DeepEqual(globalRegistryFactory.globalRegistry, noopRegistry{}) {
-		for _, c := range cs {
-			globalRegistryFactory.mustRegisterQueue = append(globalRegistryFactory.mustRegisterQueue, c)
-		}
-		return
-	}
-	globalRegistryFactory.globalRegistry.MustRegister(cs...)
-	return
+// RawRegister registers a prometheus collector but uses the global registry, this
+// bypasses the metric stability framework
+//
+// Deprecated
+func RawRegister(c prometheus.Collector) error {
```

Comment:
why do we add a deprecated func?

Reply:
Because we eventually do not want to be registering native prometheus Collectors at all, since they cannot be annotated with stability information. We require them at the moment due to the pervasive use of custom prometheus collectors in the kubelet, though we do not want to encourage or increase usage of this method going forward. @serathius is working on additional static analysis checks to aid that effort.

Reply:
good call for the immediate deprecation :+1: Can we add some explicit sentence in the godoc that this is not meant to be used for new code and point them to an alternative, i.e. native metrics? deprecated methods do tend to be used in newly added code.